service/frontend/service.go

```diff
@@ -131,8 +133,36 @@ func (s *Service) Start() {
 	// Base handler
 	s.handler = api.NewWorkflowHandler(s, s.config, client.NewVersionChecker(), dh)
 
+	userRateLimiter := quotas.NewMultiStageRateLimiter(
+		quotas.NewDynamicRateLimiter(s.config.UserRPS.AsFloat64()),
+		quotas.NewCollection(quotas.NewPerMemberDynamicRateLimiterFactory(
+			service.Frontend,
+			s.config.GlobalDomainUserRPS,
+			s.config.MaxDomainUserRPSPerInstance,
+			s.GetMembershipResolver(),
+		)),
+	)
+	workerRateLimiter := quotas.NewMultiStageRateLimiter(
+		quotas.NewDynamicRateLimiter(s.config.WorkerRPS.AsFloat64()),
+		quotas.NewCollection(quotas.NewPerMemberDynamicRateLimiterFactory(
+			service.Frontend,
+			s.config.GlobalDomainWorkerRPS,
+			s.config.MaxDomainWorkerRPSPerInstance,
+			s.GetMembershipResolver(),
+		)),
+	)
+	visibilityRateLimiter := quotas.NewMultiStageRateLimiter(
+		quotas.NewDynamicRateLimiter(s.config.VisibilityRPS.AsFloat64()),
+		quotas.NewCollection(quotas.NewPerMemberDynamicRateLimiterFactory(
+			service.Frontend,
+			s.config.GlobalDomainVisibilityRPS,
+			s.config.MaxDomainVisibilityRPSPerInstance,
+			s.GetMembershipResolver(),
+		)),
+	)
 	// Additional decorations
 	var handler api.Handler = s.handler
+	handler = ratelimited.NewAPIHandler(handler, s.GetDomainCache(), userRateLimiter, workerRateLimiter, visibilityRateLimiter)
 	handler = metered.NewAPIHandler(handler, s.GetLogger(), s.GetMetricsClient(), s.GetDomainCache(), s.config)
```

Comment:
I think I approve of this change, just making it visible / as a reminder for myself for later: this will mean that ratelimited requests no longer contribute to response latency / success rate.